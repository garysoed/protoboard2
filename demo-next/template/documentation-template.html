<style>
  #content::slotted(*) {
    padding-left: 1rem;
    padding-right: 1rem;
  }

  #title {
    padding-left: 1rem;
    padding-right: 1rem;
  }
</style>

<section flex>
  <h3 id="title" mk-display></h3>
  <slot id="content" name="content"></slot>
</section>
