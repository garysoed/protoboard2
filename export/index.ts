export {Lens, $lens} from '../src/core/lens';
export {LensDisplay} from '../src/core/lens-display';

// action
export {pickAction} from '../src/action/pick-action';

// coordinate
export {Cartesian} from '../src/coordinate/cartesian';
export {Coordinate} from '../src/coordinate/coordinate';
export {Indexed} from '../src/coordinate/indexed';

// region
export {Deck, $deck, DeckSpec} from '../src/region/deck';
export {Slot, $slot, SlotSpec, slotSpec} from '../src/region/slot';

// core
export {$createSpecEntries, $createSpecMap} from '../src/objects/object-create-spec';
export {ContentSpec} from '../src/payload/is-container';
export {ObjectSpec} from '../src/types/object-spec';
export {$render, Render} from '../src/objects/render';
export {$active, Active, activeSpec} from '../src/core/active';
export {$$activeSpec} from '../src/objects/active-spec';

// piece
export {$d1, D1, D1Spec, d1Spec} from '../src/piece/d1';
export {$d2, D2, D2Spec} from '../src/piece/d2';
export {$d6, D6, D6Spec} from '../src/piece/d6';

// tiling
export {rectOrthogonal, RectOrthogonal, Direction as RectOrthogonalDirection, distance as rectOrthogonalDistance} from '../src/tiling/rect-orthogonal';
export {vhex, VHex, Direction as VHexDirection, distance as vhexDistance} from '../src/tiling/vhex';
export {Tile} from '../src/tiling/types';
