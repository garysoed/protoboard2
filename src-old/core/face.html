<style id="style"></style>

<slot id="slot"></slot>
