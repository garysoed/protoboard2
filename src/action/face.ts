import { attributeIn, attributeOut, integerParser } from 'persona';

export const $face = {
  currentFaceIn: attributeIn('current-face', integerParser()),
  currentFaceOut: attributeOut('current-face', integerParser()),
};
