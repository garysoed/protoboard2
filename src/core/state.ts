import { Observable } from 'rxjs';

/**
 * State of a game during its runtime.
 *
 * @thModule core
 */
export interface State {
  /**
   * Identifies the object. Must be globally unique.
   */
  readonly id: string;

  /**
   * Used to map to a function that generates the object.
   */
  readonly type: string;

  /**
   * Used to generate the object.
   */
  readonly payload: ReadonlyMap<string, Observable<unknown>>;
}
