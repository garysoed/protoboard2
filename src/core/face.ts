import { Vine } from 'grapevine';
import { InstanceofType } from 'gs-types';
import { _p, stringParser, ThemedCustomElementCtrl } from 'mask';
import { attributeIn, attributeOut, element } from 'persona';

import template from './face.html';


const $$ = {
  display: attributeIn('display', stringParser()),
};

const $ = {
  host: element($$),
  slot: element('slot', InstanceofType(HTMLSlotElement), {
    name: attributeOut('name', stringParser()),
  }),
};

/**
 * A face allows you to control what to display.
 *
 * A face can have several display elements. To add a display element, add an element as a child of
 * pb-face. Each of the children must have a slot attribute with a name. This element will display
 * the display element iff the value of attribute display matches the slot name.
 */
@_p.customElement({
  tag: 'pb-face',
  template,
})
export class Face extends ThemedCustomElementCtrl {
  private readonly display$ = this.declareInput($.host._.display);

  constructor(shadowRoot: ShadowRoot, vine: Vine) {
    super(shadowRoot, vine);

    this.render($.slot._.name).withObservable(this.display$);
  }
}
