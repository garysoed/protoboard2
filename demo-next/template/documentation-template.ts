import {cache} from 'gs-tools/export/data';
import {mapNullableTo} from 'gs-tools/export/rxjs';
import {renderTheme} from 'mask';
import {Context, Ctrl, H3, iattr, id, otext, registerCustomElement} from 'persona';
import {Observable} from 'rxjs';

import template from './documentation-template.html';


const $documentationTemplate = {
  host: {
    label: iattr('label'),
  },
  shadow: {
    title: id('title', H3, {
      text: otext(),
    }),
  },
};

class DocumentationTemplate implements Ctrl {
  constructor(private readonly $: Context<typeof $documentationTemplate>) { }

  @cache()
  get runs(): ReadonlyArray<Observable<unknown>> {
    return [
      renderTheme(this.$),
      this.$.host.label.pipe(mapNullableTo(''), this.$.shadow.title.text()),
    ];
  }
}

export const DOCUMENTATION_TEMPLATE = registerCustomElement({
  ctrl: DocumentationTemplate,
  spec: $documentationTemplate,
  tag: 'pbd-documentation-template',
  template,
});
