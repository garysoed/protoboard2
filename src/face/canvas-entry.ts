import {StateId} from 'gs-tools/export/state';

export interface CanvasLine {
  readonly fromX: number;
  readonly fromY: number;
  readonly toX: number;
  readonly toY: number;
  readonly configName: string;
}

export interface CanvasHalfLine {
  readonly fromX: number;
  readonly fromY: number;
  readonly configName: string;
}

export interface CanvasIcon {
  readonly x: number;
  readonly y: number;
  readonly width: number;
  readonly height: number;
  readonly configName: string;
}


export interface CanvasEntry {
  readonly icons: StateId<readonly CanvasIcon[]>;
  readonly lines: StateId<readonly CanvasLine[]>;
  readonly halfLine: StateId<CanvasHalfLine|null>;
}
