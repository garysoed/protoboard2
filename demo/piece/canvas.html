<style>
  .slotContainer {
    display: inline-block;
    position: relative;
    width: 12rem;
    height: 12rem;
  }
</style>

<pbd-documentation-template mk-body-1 label="Canvas">
  <div slot="content">
    <p>
      Canvas lets you to draw arbitrary lines using nodes. To make the nodes very customizable,
      Protoboard doesn't come with any built in nodes.
    </p>
    <div id="canvasContainer">
      <pb-canvas></pb-canvas>
      <pbd-canvas-node></pbd-canvas-node>
      <pbd-canvas-node></pbd-canvas-node>
      <pbd-canvas-node></pbd-canvas-node>
      <pbd-canvas-node></pbd-canvas-node>
    </div>
  </div>
</pbd-documentation-template>
