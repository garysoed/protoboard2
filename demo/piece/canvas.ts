import {renderSvg} from 'almagest';
import {$stateService, source} from 'grapevine';
import {cache} from 'gs-tools/export/data';
import {StateId} from 'gs-tools/export/state';
import {BaseThemedCtrl, _p} from 'mask';
import {$div, element, PersonaContext, renderNode, RenderSpec, single} from 'persona';
import {Observable, of} from 'rxjs';
import {map} from 'rxjs/operators';

import {$canvas, Canvas, canvasSpec} from '../../src/face/canvas';
import {CanvasEntry} from '../../src/face/canvas-entry';
import canvasBgSvg from '../asset/canvas_bg.svg';
import {DocumentationTemplate} from '../template/documentation-template';

import {$canvasNode, CanvasNode} from './canvas-node';
import template from './canvas.html';

type State = CanvasEntry;

const $stateId = source(vine => $stateService.get(vine).modify(x => x.add(canvasSpec({}, x))));

export const $canvasDemo = {
  tag: 'pbd-canvas',
  api: {},
};

const $ = {
  background: element('background', $div, {
    content: single('#content'),
  }),
  canvas: element('canvas', $canvas, {}),
  circleNode: element('circleNode', $canvasNode, {}),
  noneNode: element('noneNode', $canvasNode, {}),
  squareNode: element('squareNode', $canvasNode, {}),
  triangleNode: element('triangleNode', $canvasNode, {}),
};

@_p.customElement({
  ...$canvasDemo,
  template,
  dependencies: [
    Canvas,
    CanvasNode,
    DocumentationTemplate,
  ],
})
export class CanvasDemo extends BaseThemedCtrl<typeof $> {
  constructor(context: PersonaContext) {
    super(context, $);
  }

  protected get renders(): ReadonlyArray<Observable<unknown>> {
    return [
      this.renderers.background.content(this.backgroundSvg$),
      this.renderers.canvas.objectId(this.objectId$),
      this.renderers.circleNode.objectId(this.objectId$),
      this.renderers.noneNode.objectId(this.objectId$),
      this.renderers.squareNode.objectId(this.objectId$),
      this.renderers.triangleNode.objectId(this.objectId$),
    ];
  }

  @cache()
  private get backgroundSvg$(): Observable<RenderSpec> {
    return renderSvg({
      type: 'template',
      content: canvasBgSvg,
    })
        .pipe(
            map(el => renderNode({
              node: el,
              id: {},
            })),
        );
  }

  @cache()
  private get objectId$(): Observable<StateId<State>> {
    return of($stateId.get(this.vine));
  }
}