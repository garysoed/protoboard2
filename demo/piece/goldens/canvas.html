<pbd-canvas><style>
  #canvasContainer {
    height: 130px;
    width: 130px;
    position: relative;
  }

  #circleNode, #noneNode, #squareNode, #triangleNode {
    position: absolute;
  }

  #circleNode {
    top: 0;
    left: 0;
  }

  #noneNode {
    bottom: 0;
    right: 0;
  }

  #squareNode {
    top: 0;
    right: 0;
  }

  #triangleNode {
    bottom: 0;
    left: 0;
  }
</style>

<pbd-documentation-template mk-body-1="" label="Canvas"><style>
  #content::slotted(*) {
    padding-left: 1rem;
    padding-right: 1rem;
  }

  #title {
    padding-left: 1rem;
    padding-right: 1rem;
  }
</style>

<section flex="">
  <h3 id="title" mk-display="">Canvas</h3>
  <slot id="content" name="content"><div slot="content">
    <p>
      Canvas lets you to draw arbitrary lines using nodes. To make the nodes very customizable,
      Protoboard doesn't come with any built in nodes.
    </p>
    <div id="canvasContainer">
      <div id="background">
        <!-- #content --><svg xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:cc="http://creativecommons.org/ns#" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:svg="http://www.w3.org/2000/svg" xmlns="http://www.w3.org/2000/svg" xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd" xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape" sodipodi:docname="canvas_bg.svg" inkscape:version="1.0 (4035a4f, 2020-05-01)" id="svg845" version="1.1" viewBox="0 0 34.395832 34.395863" height="130.00011" width="130">
  <defs id="defs839"></defs>
  <sodipodi:namedview units="px" inkscape:window-maximized="0" inkscape:window-y="25" inkscape:window-x="0" inkscape:window-height="997" inkscape:window-width="1680" fit-margin-bottom="0" fit-margin-right="0" fit-margin-left="0" fit-margin-top="0" showgrid="false" inkscape:document-rotation="0" inkscape:current-layer="layer1" inkscape:document-units="mm" inkscape:cy="89.619814" inkscape:cx="70.811689" inkscape:zoom="2.5083109" inkscape:pageshadow="2" inkscape:pageopacity="0.0" borderopacity="1.0" bordercolor="#666666" pagecolor="#ffffff" id="base"></sodipodi:namedview>
  <metadata id="metadata842">
    <rdf:RDF>
      <cc:Work rdf:about="">
        <dc:format>image/svg+xml</dc:format>
        <dc:type rdf:resource="http://purl.org/dc/dcmitype/StillImage"></dc:type>
        <dc:title></dc:title>
      </cc:Work>
    </rdf:RDF>
  </metadata>
  <g transform="translate(0,2.8294168e-5)" id="layer1" inkscape:groupmode="layer" inkscape:label="Layer 1">
    <path transform="scale(0.26458333)" d="M 14.886719,0 A 15,15 0 0 0 0,15 15,15 0 0 0 10,29.140625 V 100.85742 A 15,15 0 0 0 0,115 15,15 0 0 0 15,130 15,15 0 0 0 29.140625,120 H 100.85742 A 15,15 0 0 0 115,130 15,15 0 0 0 130,115 15,15 0 0 0 120,100.85938 V 29.142578 A 15,15 0 0 0 130,15 15,15 0 0 0 115,0 15,15 0 0 0 100.85938,10 H 29.142578 A 15,15 0 0 0 15,0 15,15 0 0 0 14.886719,0 Z M 29.140625,20 H 100.85742 A 15,15 0 0 0 110,29.140625 V 100.85742 A 15,15 0 0 0 100.85938,110 H 29.142578 A 15,15 0 0 0 20,100.85938 V 29.142578 A 15,15 0 0 0 29.140625,20 Z" style="fill:#666666;stroke-width:1.5219" id="path1408"></path>
  </g>
</svg>
      </div>
      <pb-canvas id="canvas" fill-parent="" object-id="4"><svg id="root" height="auto" width="auto">
  <!-- #permanents -->
  <!-- #halfline -->
</svg><link rel="stylesheet" href="base/test.css"></pb-canvas>
      <pbd-canvas-node id="squareNode" icon-type="canvas_square" x="100" y="0" object-id="4"><style>
  #root {
    background-color: transparent;
    border: 1px solid var(--mkThemeFG);
    height: 30px;
    width: 30px;
  }

  #root:hover {
    background-color: var(--mkThemeBG);
  }
</style>

<div id="root" mk-action-2=""></div><link rel="stylesheet" href="base/test.css"></pbd-canvas-node>
      <pbd-canvas-node id="circleNode" icon-type="canvas_circle" x="0" y="0" object-id="4"><style>
  #root {
    background-color: transparent;
    border: 1px solid var(--mkThemeFG);
    height: 30px;
    width: 30px;
  }

  #root:hover {
    background-color: var(--mkThemeBG);
  }
</style>

<div id="root" mk-action-2=""></div><link rel="stylesheet" href="base/test.css"></pbd-canvas-node>
      <pbd-canvas-node id="triangleNode" icon-type="canvas_triangle" x="0" y="100" object-id="4"><style>
  #root {
    background-color: transparent;
    border: 1px solid var(--mkThemeFG);
    height: 30px;
    width: 30px;
  }

  #root:hover {
    background-color: var(--mkThemeBG);
  }
</style>

<div id="root" mk-action-2=""></div><link rel="stylesheet" href="base/test.css"></pbd-canvas-node>
      <pbd-canvas-node id="noneNode" x="100" y="100" object-id="4"><style>
  #root {
    background-color: transparent;
    border: 1px solid var(--mkThemeFG);
    height: 30px;
    width: 30px;
  }

  #root:hover {
    background-color: var(--mkThemeBG);
  }
</style>

<div id="root" mk-action-2=""></div><link rel="stylesheet" href="base/test.css"></pbd-canvas-node>
    </div>
  </div></slot>
</section>
<link rel="stylesheet" href="base/test.css"></pbd-documentation-template>
<link rel="stylesheet" href="base/test.css"></pbd-canvas>